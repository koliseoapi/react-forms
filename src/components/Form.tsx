import React, {
  FieldsetHTMLAttributes,
  FormEvent,
  FormHTMLAttributes,
  ReactEventHandler,
  createContext,
  useContext,
  useEffect,
  useRef,
  useState,
} from "react";
import { Messages } from "../core/Messages";
import { getNestedProperty, setNestedProperty } from "../core/utils";
import { ValidationResult } from "../core/ValidationActions";
import { Validator, Validators } from "../core/Validators";
import { I18nContext } from "./I18nContext";
import { BoundComponentProps } from "./InputElements";

interface FormProps extends FormHTMLAttributes<HTMLFormElement> {
  /** the object being edited  */
  initialValues: any;

  /** the method to invoke when submitting, after passing all validations */
  onSubmit: (values: any) => Promise<any>;
}

/**
 * The list of all validation errors. The key is the propertyName,
 * the value is the error message
 */
export interface ValidationErrors {
  [propertyName: string]: string;
}

/**
 * Context to share with child BoundComponents
 */
export interface FormContextContent {
  /**
   * All error messages, indexed by property name
   */
  errors: ValidationErrors;

  /**
   * True if this form is submitting the value
   */
  submitting: boolean;

  /**
   * Add a Validator to the list of fields validated by this form
   */
  addValidator(props: BoundComponentProps): void;

  /**
   * Remove a validator
   */
  removeValidator(name: string): void;

  /**
   * Triggered by a change in the form field
   */
  setValue(propertyName: string, value: any): void;

  /**
   * Get value associated to a BoundComponent
   */
  getValue(propertyName: string): any;

  submit(): void;
}

function undefinedFormContext(): never {
  throw new Error("A <Form> context is required");
}
export const FormContext = createContext<FormContextContent>({
  addValidator: undefinedFormContext,
  removeValidator: undefinedFormContext,
  setValue: undefinedFormContext,
  getValue: undefinedFormContext,
  errors: {},
  submitting: false,
  submit() {},
});

/**
 * Form is the required container  of BoundComponent instances.
 */
export function Form({
  initialValues,
  onSubmit,
  children,
  ...props
}: FormProps) {
  // errors to display, sorted by component key
  const [errors, setErrors] = useState<ValidationErrors>({});

  // object with the data introduced in this form
  const [values, setValues] = useState({ ...initialValues });

  // the list of validators to use for this object
  const [validators, setValidators] = useState<Validators>({});

  // the submission status
  const [submitting, setSubmitting] = useState(false);

  // should focus on the first component with error
  const [focusFirstError, setFocusFirstError] = useState(false);

  // control if this form is still mounted
  const mounted = useRef(true);

  // validation rules to apply
  const entries = useContext(I18nContext);
  const messages = new Messages(entries);

  const onSubmitHandler: ReactEventHandler = async function (e): Promise<any> {
    e.preventDefault();
    e.stopPropagation();
    const errors = await validate();
    if (!Object.keys(errors).length) {
      setSubmitting(true);
      try {
        await onSubmit(values);
      } finally {
        // could be that the form is already dismounted
        mounted.current && setSubmitting(false);
      }
    } else {
      return errors;
    }
  };

  const formContextValue: FormContextContent = {
    errors,
    submitting,
    removeValidator: function (name) {
      delete validators[name];
      setValidators(validators);
    },

    addValidator: function (props) {
      validators[props.name] = new Validator(props);
      setValidators(validators);
    },

    setValue: function (propertyName, value) {
      setNestedProperty(values, propertyName, value);
      if (errors[propertyName]) {
        delete errors[propertyName];
        setErrors({ ...errors });
      }
      setValues(values);
    },

    getValue(propertyName) {
      return getNestedProperty(values, propertyName);
    },

    submit() {
      onSubmitHandler({
        preventDefault() {},
        stopPropagation() {},
      } as FormEvent);
    },
  };

  async function validate(): Promise<ValidationErrors> {
    const newErrors: ValidationErrors = {};
    await Promise.all(
      Object.entries(validators).map(async ([propertyName, validator]) => {
        const error: ValidationResult = await validator.validate(
          getNestedProperty(values, propertyName)
        );
        if (!!error) {
          newErrors[propertyName] = messages.get(error, validator.props);
        }
      })
    );
    setErrors(newErrors);
    if (Object.keys(newErrors).length > 0) {
      setFocusFirstError(true);
    }
    return newErrors;
  }

  useEffect(() => {
    if (focusFirstError) {
      const firstInvalidElement = document.body.querySelector<HTMLInputElement>(
        "[aria-invalid=true]"
      );
      firstInvalidElement && firstInvalidElement.focus();
      setFocusFirstError(false);
    }
  }, [focusFirstError]);

  useEffect(() => {
    // React 18 in dev mode with strict mode seems to run useEffect twice
    // https://github.com/facebook/react/issues/24670#issuecomment-1280222217
    mounted.current = true;
    return () => {
      mounted.current = false;
    };
  }, []);

  return (
    <form {...props} onSubmit={onSubmitHandler} noValidate>
      <FormContext.Provider value={formContextValue}>
        {children}
      </FormContext.Provider>
    </form>
  );
}

/**
 * A fieldset that will disable all input controls while submitting
 */
export function FieldSet({
  children,
  ...props
}: FieldsetHTMLAttributes<HTMLFieldSetElement>) {
  const formContext = useContext(FormContext);
  return (
    <fieldset disabled={formContext.submitting} {...props}>
      {children}
    </fieldset>
  );
}
