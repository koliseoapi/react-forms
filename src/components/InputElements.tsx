import React, {
  useContext,
  ChangeEvent,
  useEffect,
  InputHTMLAttributes,
  useRef,
} from "react";
import { Converters, Converter } from "../core/Converters";
import { FormContext, FormContextContent } from "./Form";
import { ValidationResult } from "../core/ValidationActions";
import { errorStyles } from "../core/utils";

export interface BoundComponentProps
  extends Omit<InputHTMLAttributes<HTMLInputElement>, "type" | "onChange"> {
  /** propertyName. Will also be used as id if none is specified */
  name: string;

  /** Optional: the converter to use. If not set, the default for the input type will be used */
  converter?: Converter<any>;

  /** Optional: the validator to use. If set, all default validators (required, number, etc) will be replaced by this. Received the value after being converted */
  validate?(value: any): Promise<ValidationResult>;

  /** onChange event handler has been extended to also receive the formContext */
  onChange?(
    event: ChangeEvent<HTMLInputElement>,
    formContext: FormContextContent
  ): void;
}

export interface InputProps extends BoundComponentProps {
  type:
    | "text"
    | "url"
    | "email"
    | "number"
    | "checkbox"
    | "date"
    | "time"
    | "radio"
    | "datetime-local";
}

export interface BoundComponentPropsWithElement extends BoundComponentProps {
  /** type of input component to use */
  elementName: "input" | "select" | "textarea";
}

/**
 * An input/select/textarea component that will bind automatically
 * with the corresponding object propety in the containing `Form`
 */
export function BoundComponent({
  elementName,
  converter,
  validate,
  name,
  id,
  onChange: originalOnChange,
  children,
  ...props
}: BoundComponentPropsWithElement) {
  const type: string = (props as any).type;
  if (process.env.NODE_ENV !== "production") {
    if (type === "radio" && typeof props.defaultValue === "undefined") {
      throw new Error(`Radio button ${name} requires a defaultValue`);
    }
    if (typeof props.value !== "undefined") {
      throw new Error(`${name} should use defaultValue instead of value`);
    }
  }

  if (!id) {
    id = name.replace(/\./g, "-");
    if (type === "radio") {
      id = `${name}_${props.defaultValue}`;
    }
  }
  converter = converter || Converters[type] || Converters.text;
  const formContext = useContext<FormContextContent>(FormContext);
  const defaultValue =
    type === "checkbox"
      ? undefined
      : type === "radio"
      ? (props.defaultValue as string)
      : converter.toValue(formContext.getValue(name));
  const defaultChecked =
    type === "checkbox"
      ? formContext.getValue(name)
      : type === "radio"
      ? formContext.getValue(name) == defaultValue
      : undefined;

  function onChange(e: ChangeEvent<HTMLInputElement>) {
    const element = e.target;
    const objectValue =
      type === "checkbox"
        ? element.checked
        : converter!.fromValue({ value: element.value, ...props });
    formContext.setValue(name, objectValue);
    originalOnChange && originalOnChange(e, formContext);
  }

  useEffect(() => {
    formContext.addValidator({ name, validate, converter, ...props });
    return () => formContext.removeValidator(name);
  }, []);

  const error = formContext.errors[name];

  // ARIA attributes to point to the error message
  const errorMessageId = error ? `${id}_error` : undefined;
  const adb = props["aria-describedby"];
  const ariaProps = !error
    ? {}
    : {
        "aria-invalid": true,
        "aria-describedby": adb ? `${errorMessageId} ${adb}` : errorMessageId,
      };
  const ref = useRef(null);

  // submit on Ctrl + Enter
  function onKeyDown(e: KeyboardEvent) {
    if (e.key === "Enter" && e.ctrlKey) {
      formContext.submit();
    }
  }

  return (
    <>
      {React.createElement(
        elementName,
        {
          name,
          id: id,
          onChange,
          defaultChecked,
          defaultValue,
          ref,
          onKeyDown,
          ...props,
          ...ariaProps,
        },
        children
      )}
      {error ? (
        <div
          className="input-error"
          id={errorMessageId}
          role="alert"
          style={errorStyles(ref)}
        >
          {error}
        </div>
      ) : undefined}
    </>
  );
}

export function Input({ autoComplete = "off", ...props }: InputProps) {
  return (
    <BoundComponent
      elementName="input"
      autoComplete={autoComplete}
      {...props}
    />
  );
}

export function Select(props: BoundComponentProps) {
  return <BoundComponent elementName="select" {...props} />;
}

export function TextArea(props: BoundComponentProps) {
  return <BoundComponent elementName="textarea" {...props} />;
}
