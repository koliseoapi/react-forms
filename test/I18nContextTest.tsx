/**
   TODO multilanguage
    it("should pass props to customize validation messages", function () {
    const validators = new ValidationActions(
      new Messages({
        required: "Please fill out ${name}",
        min: "${name} value must be greater than or equal to ${min}",
        max: "Value must be less than or equal to ${max}",
        url: "Please enter a URL for ${name}",
        email: "Please include a valid e-mail address for ${name}",
        pattern: "Please match the requested format for ${name}",
        maxLength: "Value must have no more than ${maxLength} characters",
      })
    );
    const props = { name: "Foo" };
    expect(ValidationActions.required(undefined, props)).toBe("Please fill out Foo");
    expect(ValidationActions.number_min(-1, { ...props, min: 0 })).toBe(
      "Foo value must be greater than or equal to 0"
    );
    expect(ValidationActions.number_max(6, { ...props, max: 5 })).toBe(
      "Value must be less than or equal to 5"
    );
    expect(ValidationActions.url("foo", props)).toBe("Please enter a URL for Foo");
    expect(ValidationActions.email("foo", props)).toBe(
      "Please include a valid e-mail address for Foo"
    );
    expect(ValidationActions.pattern("foo", { ...props, pattern: "[0-9]+" })).toBe(
      "Please match the requested format for Foo"
    );
    expect(ValidationActions.maxLength("foo", { ...props, maxLength: 2 })).toBe(
      "Value must have no more than 2 characters"
    );
  });

 */

import React from "react";
import renderer, { act, ReactTestRenderer } from "react-test-renderer";
import { Form } from "../src/components/Form";
import { I18nContext } from "../src/components/I18nContext";
import { Input } from "../src/components/InputElements";

describe("Input", function () {
  let onSubmit: () => Promise<string>;

  let form: ReactTestRenderer;

  it("should render a validation error", async function () {
    onSubmit = jest.fn(async () => "foo");
    // act() to wait for the useEffect() calls on initial render
    act(() => {
      form = renderer.create(
        <I18nContext.Provider
          value={{ required: "Por favor, introduzca un valor" }}
        >
          <Form onSubmit={onSubmit} initialValues={{}}>
            <Input type="text" required name="name" />
          </Form>
        </I18nContext.Provider>
      );
    });
    await act(async () => {
      const formElement = form.root.find((el) => el.type == "form");
      try {
        await formElement.props.onSubmit({
          preventDefault() {},
          stopPropagation() {},
        });
      } catch (errors) {
        expect(errors).toMatchObject({
          name: "Por favor, introduzca un valor",
        });
      }
    });
    expect(form.toJSON()).toMatchSnapshot();
  });
});
