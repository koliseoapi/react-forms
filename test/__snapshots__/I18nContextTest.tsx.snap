// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render a validation error 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <input
    aria-describedby="name_error"
    aria-invalid={true}
    autoComplete="off"
    defaultValue=""
    id="name"
    name="name"
    onChange={[Function]}
    onKeyDown={[Function]}
    required={true}
    type="text"
  />
  <div
    className="input-error"
    id="name_error"
    role="alert"
  >
    Por favor, introduzca un valor
  </div>
</form>
`;
